### PR TITLE
Add IM Event Lock support

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Need to implement IM event lock

#### Change overview
Use mutex from SystemMutex.h, and remove unused CriticalSection code.

#### Testing
The existing unit test has covered this.

### CODE REVIEW

File: src/app/EventManagement.cpp [hunk 1]
```diff
@@ -33,6 +33,26 @@ namespace chip {
...
+
+class EventLocked
+{
+public:
+    EventLocked() {}
+    ~EventLocked() {}
+};
+
+#else
+System::Mutex sEventLock;
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:46]: Why is this not a class static in the EventLocked class, instead of being a global?

File: src/app/EventManagement.cpp [hunk 2]
```diff
@@ -33,6 +33,26 @@ namespace chip {
...
+
+class EventLocked
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:48]: Maybe `LockEvents`?

File: src/app/EventManagement.cpp [hunk 3]
```diff
@@ -462,15 +481,14 @@ CHIP_ERROR EventManagement::CopyAndAdjustDeltaTime(const TLVReader & aReader, si
...
-    CriticalSectionEnter();
+    EventLocked lock;
 
     VerifyOrExit(mState != EventManagementStates::Shutdown, err = CHIP_ERROR_INCORRECT_STATE);
     VerifyOrExit(aEventOptions.mpEventSchema != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
 
     err = LogEventPrivate(apDelegate, aEventOptions, aEventNumber);
 
 exit:
-    CriticalSectionExit();
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:473]: Do we want to be holding the lock when we call `ChipLogFunctError`?  Might be better to have a narrower scope for the lock here.

File: src/app/EventManagement.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: Done, thanks

File: src/app/EventManagement.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:473]: Agree, done, thanks

File: src/app/EventManagement.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: add static, done, thanks

**Review by bzbarsky-apple — COMMENTED** [review]
Why the change back away from RAII?

File: src/app/EventManagement.cpp [hunk 4]
```diff
@@ -33,6 +33,21 @@ namespace chip {
 namespace app {
 static EventManagement sInstance;
 
+#if CHIP_SYSTEM_CONFIG_NO_LOCKING
```
Comment (mrjerryjohns) [inline src/app/EventManagement.cpp:36]: I'm confused by why we're using this macro to conditionally enable/disable the use of scoped locks?

File: src/app/EventManagement.cpp [hunk 5]
```diff
@@ -33,6 +33,21 @@ namespace chip {
...
+static void LockEvents() {}
+static void UnLockEvents() {}
+#else
+static System::Mutex sEventLock;
+static void LockEvents()
```
Comment (mrjerryjohns) [inline src/app/EventManagement.cpp:41]: EventManagement is already a global singleton, so it'd be cleaner to add LockEvents() as a public API method on that object.

Moreover, if we plan to use any #ifdef treatment here, the actual ifdef should be moved within the LockEvent() implementation itself, not outside it.

Comment (mrjerryjohns) [reply src/app/EventManagement.cpp:41]: If you move it in, then the "Events" part of the "LockEvents" is kind of redundant.

Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: Done, thanks

File: src/app/EventManagement.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: Since this scope is originated from SystemMutex.cpp/h, I think it is for new product/os where system lock has not yet implemented

**Review by yunhanw-google — COMMENTED** [review]
> Why the change back away from RAII?

Follow Jerry's suggestion and move it inside EventManagement class, without RAII, we can unlock before we call ChipLogFunctError

File: src/app/EventManagement.cpp [hunk 6]
```diff
@@ -25,13 +24,16 @@
...
 #include <support/logging/CHIPLogging.h>
+#include <system/SystemMutex.h>
 #include <system/SystemTimer.h>
 
 using namespace chip::TLV;
 
 namespace chip {
 namespace app {
 static EventManagement sInstance;
+static System::Mutex sEventLock;
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:35]: Why is this not a member of EventManagement?  Might be worth documenting that, at least....

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/EventManagement.cpp (hunk 6 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: forgot to move it, just update it as member variable inside EventManagement.

File: src/app/EventManagement.h [hunk 7]
```diff
@@ -566,6 +560,7 @@ class EventManagement
     Messaging::ExchangeManager * mpExchangeMgr = nullptr;
     EventManagementStates mState               = EventManagementStates::Shutdown;
     uint32_t mBytesWritten                     = 0;
+    static System::Mutex mEventLock;
```
Comment (andy31415) [inline src/app/EventManagement.h:563]: When reading the header, I am wondering what 'event lock' protects. I would have expected a 'Event static member' but that is not the case:

- Why is this static?
- can we name it like "EventBufferLock" if it is supposed to the event buffer?

File: src/app/EventManagement.h [hunk 8]
```diff
@@ -551,6 +542,9 @@ class EventManagement
      */
     static CHIP_ERROR CopyEvent(const TLV::TLVReader & aReader, TLV::TLVWriter & aWriter, EventLoadOutContext * apContext);
 
+    static void Lock();
```
Comment (andy31415) [inline src/app/EventManagement.h:545]: why static?

File: src/app/EventManagement.h (hunk 7 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h]: just remove static, previously there is check static_assert(std::is_trivially_destructible<EventManagement>::value, "EventManagement must be trivially destructible"); Currently System::Mutex is not trivally destructible, I think it is ok to remove this static_assert

File: src/app/EventManagement.h (hunk 8 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h]: just remove static

**Comment by msandstedt** [comment]
> > Why the change back away from RAII?
> 
> Follow Jerry's suggestion and move it inside EventManagement class, without RAII, we can unlock before we call ChipLogFunctError

Wait, is this actually true?  Can't the following work?

```
CHIP_ERROR EventManagement::LogEvent(EventLoggingDelegate * apDelegate, EventOptions & aEventOptions, EventNumber & aEventNumber)
{
    CHIP_ERROR err = CHIP_NO_ERROR;
    {
        RAIILock raii_lock(&lock);
        VerifyOrExit(mState != EventManagementStates::Shutdown, err = CHIP_ERROR_INCORRECT_STATE);
        VerifyOrExit(aEventOptions.mpEventSchema != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
        err = LogEventPrivate(apDelegate, aEventOptions, aEventNumber);
    }

exit:
    ChipLogFunctError(err);
    return err;
}

```

**Comment by mrjerryjohns** [comment]
@msandstedt I was just about to suggest the same. An RAII lock scheme would be better....

**Comment by yunhanw-google** [comment]
Just move back to RAII, thanks

File: src/app/EventManagement.cpp (hunk 3 above)
Comment (msandstedt) [reply src/app/EventManagement.cpp:473]: This is now back to holding the lock through the `ChipLogFunctError` call.

With an RAII lock, the duration of locking is controlled by scope.  Here, scope can be explicitly introduced with braces:

```
CHIP_ERROR EventManagement::LogEvent(EventLoggingDelegate * apDelegate, EventOptions & aEventOptions, EventNumber & aEventNumber)
{
    CHIP_ERROR err = CHIP_NO_ERROR;
    {
        LockEvents lock;

        VerifyOrExit(mState != EventManagementStates::Shutdown, err = CHIP_ERROR_INCORRECT_STATE);
        VerifyOrExit(aEventOptions.mpEventSchema != nullptr, err = CHIP_ERROR_INCORRECT_STATE);

        err = LogEventPrivate(apDelegate, aEventOptions, aEventNumber);
    } // lock is released here by implicit destruction

exit:
    ChipLogFunctError(err);
    return err;
}
```

Is there a reason this can't work?  Does the `goto` in `VerifyOrExit` somehow prevent it?

File: src/inet/InetLayer.h [hunk 9]
```diff
@@ -92,7 +92,11 @@
 #endif // CHIP_SYSTEM_CONFIG_POSIX_LOCKING
 
 #if CHIP_SYSTEM_CONFIG_FREERTOS_LOCKING
+#if defined(ESP_PLATFORM)
+#include "freertos/FreeRTOS.h"
+#else
```
Comment (tcarmelveilleux) [inline src/inet/InetLayer.h:97]: Why is this needed in InetLayer.h?

File: src/system/SystemMutex.h [hunk 10]
```diff
@@ -78,8 +84,8 @@ class DLL_EXPORT Mutex
 #if (configSUPPORT_STATIC_ALLOCATION == 1)
     StaticSemaphore_t mFreeRTOSSemaphoreObj;
 #endif // (configSUPPORT_STATIC_ALLOCATION == 1)
-    volatile SemaphoreHandle_t mFreeRTOSSemaphore;
-    volatile int mInitialized;
+    volatile SemaphoreHandle_t mFreeRTOSSemaphore = nullptr;
```
Comment (tcarmelveilleux) [inline src/system/SystemMutex.h:87]: Why not only ever use static semaphores on FreeRTOS?

Also, what about Zephyr?

File: src/app/EventManagement.cpp [hunk 11]
```diff
@@ -462,15 +477,14 @@ CHIP_ERROR EventManagement::CopyAndAdjustDeltaTime(const TLVReader & aReader, si
...
+    LockEvents lock;
 
     VerifyOrExit(mState != EventManagementStates::Shutdown, err = CHIP_ERROR_INCORRECT_STATE);
     VerifyOrExit(aEventOptions.mpEventSchema != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
 
     err = LogEventPrivate(apDelegate, aEventOptions, aEventNumber);
 
 exit:
-    CriticalSectionExit();
     ChipLogFunctError(err);
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:488]: As @msandstedt points out this now ends up happening while locked.  We should not do that.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/EventManagement.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:473]: Done, thanks

File: src/app/EventManagement.h [hunk 12]
```diff
@@ -551,6 +541,9 @@ class EventManagement
      */
     static CHIP_ERROR CopyEvent(const TLV::TLVReader & aReader, TLV::TLVWriter & aWriter, EventLoadOutContext * apContext);
 
+    void Lock();
+    void Unlock();
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/EventManagement.h]: removed, thanks

File: src/app/EventManagement.cpp (hunk 11 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:470]: Updated per @msandstedt 's suggestion.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/system/SystemMutex.h (hunk 10 above)
Comment (yunhanw-google) [reply src/system/SystemMutex.h]: if using StaticSemaphore_t, in esp32, I am seeing invalid cast from type 'StaticSemaphore_t' {aka 'xSTATIC_QUEUE'} to type 'QueueHandle_t' {aka 'void*'}
     xSemaphoreGive(this->mFreeRTOSSemaphore)，
in espe32, i see typedefQueueHandle_tSemaphoreHandle_t

File: src/inet/InetLayer.h (hunk 9 above)
Comment (yunhanw-google) [reply src/inet/InetLayer.h:97]: in InetLayer.h, I am seeing 
#elif CHIP_SYSTEM_CONFIG_FREERTOS_LOCKING
#if (configSUPPORT_STATIC_ALLOCATION == 1)
    StaticSemaphore_t mDroppableEventsObj;
#endif // (configSUPPORT_STATIC_ALLOCATION == 1)
    SemaphoreHandle_t mDroppableEvents;
#endif // CHIP_SYSTEM_CONFIG_FREERTOS_LOCKING

Wihthout this #include "freertos/FreeRTOS.h", compilation fails in esp32.

File: src/app/EventManagement.h [hunk 13]
```diff
@@ -298,6 +289,25 @@ class EventManagement
 
     static void DestroyEventManagement();
 
+    class LockEvents
+    {
+    public:
+        LockEvents(EventManagement & aEventManagement) : mEventManagement(aEventManagement)
+        {
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
```
Comment (tcarmelveilleux) [inline src/app/EventManagement.h:297]: This whole CHIP_SYSTEM_CONFIG_NO_LOCKING define seems odd here. You cannot have something called a Lock in code that you can't tell unless you look at the implementation that it does nothing. Which platforms have this define and why? Could you call it ProtectEvents instead? Where are the side effects of CHIP_SYSTEM_CONFIG_NO_LOCKING being used called out?
.

File: src/app/EventManagement.h [hunk 14]
```diff
@@ -566,6 +576,7 @@ class EventManagement
     Messaging::ExchangeManager * mpExchangeMgr = nullptr;
     EventManagementStates mState               = EventManagementStates::Shutdown;
     uint32_t mBytesWritten                     = 0;
+    System::Mutex mAccessLock;
```
Comment (tcarmelveilleux) [inline src/app/EventManagement.h:579]: Does the mutex act as a noop without CHIP_SYSTEM_CONFIG_NO_LOCKING defined?

File: src/system/SystemMutex.cpp [hunk 15]
```diff
@@ -82,7 +82,7 @@ DLL_EXPORT Error Mutex::Init(Mutex & aThis)
 #else
         aThis.mFreeRTOSSemaphore = xSemaphoreCreateMutex();
 #endif
-        if (aThis.mFreeRTOSSemaphore == NULL)
+        if (aThis.mFreeRTOSSemaphore == nullptr)
```
Comment (tcarmelveilleux) [inline src/system/SystemMutex.cpp:85]: All of this weird atomic code is only needed if mutex init is called from many threads at once, which would be an error...

File: src/system/SystemMutex.h (hunk 10 above)
Comment (tcarmelveilleux) [reply src/system/SystemMutex.h:87]: Why not try to resolve it rather than having two implementations for the same RTOS. I don't recall offhand, but I think static semaphore type is always supported. @dhrishi for ESP feedback.

Overall, instead of many #ifdefs in core code, why isn't this SystemMutex abstraction just implemented per platform as opposed to once non-portably withany ifdefs?

File: src/app/EventManagement.h [hunk 16]
```diff
@@ -298,6 +289,25 @@ class EventManagement
 
     static void DestroyEventManagement();
 
+    class LockEvents
```
Comment (andy31415) [inline src/app/EventManagement.h:292]: s/LockEvents/ScopedLock

File: src/system/SystemMutex.h (hunk 10 above)
Comment (andy31415) [reply src/system/SystemMutex.h:87]: Fixing semaphores should probably be a separate PR

File: src/app/EventManagement.h (hunk 13 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h:297]: just rename it to ScopedLock, we have enable lock for all supported platform, so it is safe to remove CHIP_SYSTEM_CONFIG_NO_LOCKING here.

File: src/app/EventManagement.h (hunk 14 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h:579]: removed, thanks

File: src/app/EventManagement.h (hunk 16 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h:292]: Done, thanks

File: src/system/SystemMutex.h (hunk 10 above)
Comment (yunhanw-google) [reply src/system/SystemMutex.h:87]: Create https://github.com/project-chip/connectedhomeip/issues/7415 for issue track

File: src/system/SystemMutex.cpp (hunk 15 above)
Comment (yunhanw-google) [reply src/system/SystemMutex.cpp:85]: https://github.com/project-chip/connectedhomeip/issues/7415
It would be resolved in seperate PR, thanks

**Review by msandstedt — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

File: src/app/EventManagement.cpp [hunk 17]
```diff
@@ -134,6 +126,8 @@ void EventManagement::Init(Messaging::ExchangeManager * apExchangeManager, int a
     mpEventBuffer = apCircularEventBuffer;
     mState        = EventManagementStates::Idle;
     mBytesWritten = 0;
+
+    chip::System::Mutex::Init(mAccessLock);
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: Done, thanks

**Review by tcarmelveilleux — APPROVED** [review]
